narrow: Make `with` operator behaviour consistent with API behaviour.

roanster007 opened the pull request:
Previously, in presence of `dm` operators, the `with` operator behaviour in the web client was not consistent with that of the API. For instance, when `with` points to a stream message, while the narrow containing `dm`, rather than correcting narrow to point to the anchor message, it used to remove the `with`.

Similarly, when `with` pointed to a direct message, then it used to narrow to the correct narrow, but `with` used to be removed from the filter.

This commit removes these inconsistencies of `with` in case of `dm` operators, and makes it consistent to those mentioned in the api documentation.

<!-- Describe your pull request here.-->
[CZO thread](https://chat.zulip.org/#narrow/stream/378-api-design/topic/topic.20permalinks/near/1907306)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented:
@Vector73 can you review it

Vector73 commented on web/src/filter.ts:
```diff
@@ -822,12 +823,14 @@ export class Filter {
             assert(typeof message.display_recipient !== "string");
             const dm_participants = message.display_recipient.map((user) => user.email);
-            const dm_operand = dm_participants.join(",");
+            const dm_operand = dm_participants
+                .filter((user_email) => user_email !== current_user.email)
+                .join(",");
```
Can you explain the reasoning behind this change? Will this work as expected if there is only `current_user.email` in `dm_participants`?

roanster007 replied:
Ya, your point makes sense! I was actually trying to fix the bug I mentioned [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/Fading.20of.20messages.20of.20different.20conversations.20in.20DM/near/1919676). I'll remove this change and try to fix it separately.

Vector73 reviewed (commented):
Sorry for the late review. I was a little busy this week, so was unable to take a look at this. I left a small comment. Otherwise, looks good to me.

roanster007 commented:
@zulipbot  remove "buddy review"
@zulipbot add "mentor review"

roanster007 commented:
@prakhar1144 can you review this?

prakhar1144 commented on web/tests/filter.test.js:
```diff
@@ -2446,10 +2454,7 @@ run_test("adjusted_terms_if_moved", () => {
-        {operator: "dm", operand: "user3@zulip.com", negated: false},
-        {operator: "with", operand: "2"},
-    ]);
+    assert.deepEqual(result, [{operator: "dm", operand: "user3@zulip.com", negated: false}]);
```
How was this working previously ? -- I don't see `with` retained in the url in this case when tested out in CZO.

roanster007 replied:
So basically it would check the message id, and since it would update the narrow accordingly, and then remove the `with` term -- same as in https://github.com/zulip/zulip/pull/31239/files#r1721784931

Previously, the `with` was removed before correcting the narrow and now the narrow is being corrected and then `with` is removed for `dm` terms

prakhar1144 reviewed (commented):
For a link like:
http://localhost:9991/#narrow/stream/11-devel/topic/desktop/with/79

message_id is a DM ID. We narrow to the DM conversation but the `with` part is lost in the url. Is this intended ?

The reverse works as expected i.e. For:
http://localhost:9991/#narrow/dm/6,13-dm/with/61

61 is stream message ID. The with part is retained.

roanster007 commented on web/src/filter.ts:
```diff
@@ -827,7 +828,7 @@ export class Filter {
             const filtered_terms = raw_terms.filter((term) => {
                 const operator = Filter.canonicalize_operator(term.operator);
-                return !conversation_terms.has(operator);
+                return !conversation_terms.has(operator) && operator !== "with";
```
>For a link like:
http://localhost:9991/#narrow/stream/11-devel/topic/desktop/with/79
message_id is a DM ID. We narrow to the DM conversation but the with part is lost in the url. Is this intended ?

@prakhar1144 ya, it was intended, I added it here -- I thought like adding `with` to a `dm` narrow at this point wouldn't make sense since dm links are already permalinks. However, if you add `with` to a `dm` narrow now, they wouldn't be removed unlike previous, and would take you to the right narrow.

roanster007 commented on web/src/filter.ts:
```diff
@@ -831,7 +832,17 @@ export class Filter {
             assert(typeof message.display_recipient !== "string");
-            const dm_participants = message.display_recipient.map((user) => user.email);
+
+            // We should make sure the current user is not included for
+            // the `dm` operand for the narrow.
+            const dm_participants = message.display_recipient
+                .map((user) => user.email)
+                .filter((user_email) => user_email !== current_user.email);
+            // However, if the current user is the only recipient of the
+            // message, we should include the user in the operand.
+            if (dm_participants.length === 0) {
+                dm_participants.push(current_user.email);
+            }
```
This is important because without this, the recipients for `dm` narrow would also contain `current_user`. This would matter when you try to get bucketed unread messages -- since the unreads are bucketed with recipients that don't include `current_user`, the unreads returned would be `[]`

roanster007 commented on web/src/filter.ts:
```diff
@@ -879,6 +890,8 @@ export class Filter {
         this.cached_sorted_terms_for_comparison = undefined;
         if (this.has_operator("channel")) {
             this._sub = stream_data.get_sub_by_name(this.operands("channel")[0]!);
+        } else {
+            this._sub = undefined;
```
This is for case when `#narrow/channel/topic/with/private_msg` -- when the narrow terms are adjusted and new filter with only dm terms is set up, the `_sub` of old filter needs to be removed.

prakhar1144 commented:
Update: This PR looks great apart from [a bugfix which deserves it's own commit / PR](https://github.com/zulip/zulip/pull/31239#discussion_r1729154049) (discussed with Rohan in DMs as I was confused about why that change is required in this commit)

roanster007 commented:
>Update: This PR looks great apart from https://github.com/zulip/zulip/pull/31239#discussion_r1729154049 (discussed with Rohan in DMs as I was confused about why that change is required in this commit)

Extracted it from this PR to #31494

prakhar1144 reviewed (changes requested):
Works as expected in manual testing.

The commit message is not quite correct:
> For instance, when `with` points to a stream message, while the narrow containing `dm`, rather than correcting narrow to point to the anchor message, it used to remove the `with`.

"*it used to remove the `with`*" -- It didn't remove the `with`, instead it just narrow us to the `dm` narrow without considering the effect of `with`.

For the second case, i.e.
> Similarly, when `with` pointed to a direct message, then it used to narrow to the correct narrow, but `with` used to be removed from the filter.

This only happens in the presense of `stream` operator and `with` operator with message ID of a DM. The current commit message says "Similarly" which means "in the presence of `dm` operator" (as said in the prev paragraph of commit message) -- which is not correct. It works as expected in main when dm + with-dm-message-id is used.

roanster007 commented:
Oh, thanks for catching those, I should have been more specific.

>"it used to remove the with" -- It didn't remove the with, instead it just narrow us to the dm narrow without considering the effect of with.

Updated to
```
...For instance, when `with` points to a stream message,
in presence of a `dm` term, then rather than narrowing to the
corresponding channel of `with` operand, the `with` term simply
used to be ignored, and narrowed to the corresponding `dm`
narrow.
```

>This only happens in the presense of stream operator and with operator with message ID of a DM. The current commit message says "Similarly" which means "in the presence of dm operator" (as said in the prev paragraph of commit message) -- which is not correct. It works as expected in main when dm + with-dm-message-id is used.

Updated to 
```
Also, in cases when the `with` used to point to a direct message,
in presence of a `channel` term, then after correcting to the
right narrow, the `with` term used to be removed. The `with`
term would still be needed after correcting the narrow to
maintain consistency between channel and DM conversations.
```

prakhar1144 reviewed (approved):
Thanks!

The code LGTM, works as expected in manual tesing.

timabbott commented on web/src/narrow_state.ts:
```diff
@@ -349,7 +349,7 @@ export function narrowed_by_pm_reply(current_filter: Filter | undefined = filter
-    const terms = current_filter.terms();
+    const terms = current_filter.terms().filter((term) => term.operator !== "with");
     return terms.length === 1 && current_filter.has_operator("dm");
 }
```
Should the `near` operator be handled here too? Overall I'm worrying that we're not processing `near/with` operators in the right places; maybe those should not be stored as `terms` on the filter object.

timabbott commented on web/src/filter.ts:
```diff
@@ -402,7 +402,8 @@ export class Filter {
         }
 
-        const updated_terms = [...orig_terms];
+        const updated_terms = orig_terms.filter((term: NarrowTerm) => term.operator !== "dm");
```
This function is kinda a mess; Is it even possible to have a `dm` operator here? 

I think this code path should be rewritten to just pull off any channel/topic terms and add the correct new ones, rather than having this temporary `""` operands approach, which just feels very fragile.

But that can probably be a follow-up PR; this new code does not seem obviously worse.

roanster007 replied:
>This function is kinda a mess; Is it even possible to have a dm operator here?

This comes up when we have a dm narrow with a with term pointing to a stream message (the dm term would be removed and channel-topic terms would be added here).

timabbott commented:
Merged, but note the two above things that I noticed (not new bugs in this code) that likely deserve attention.

Thanks @roanster007!
